File: archive/2019/solutions/ContradictoryCanele.cpp
#include <bits/stdc++.h>

using namespace std;

long long grid[205][205];

int main() {
  long long n, m, k, q;
  cin >> n >> m >> k >> q;
  if (n % k == 0 && m % q == 0) {
    cout << "NO" << endl;
    exit(0);
  }
  cout << "YES" << endl;
  for (int i = 0; i < k; i++) {
    for (int j = 0; j < q; j++) {
      grid[i][j] = k * q - (i * q + j) - 1;
    }
  }
  grid[k - 1][q - 1] = -(k * q) * (k * q - 1) / 2;
  for (int i = 0; i < n; i++) {
    for (int j = 0; j < m; j++) {
      cout << -grid[i % k][j % q] << " ";
    }
    cout << endl;
  }
}

File: archive/2019/solutions/ContradictoryCanele.py
def main():
    grid = []

    ans = ""

    stin = input().split()
    n = int(stin[0])
    m = int(stin[1])
    k = int(stin[2])
    q = int(stin[3])

    if n % k == 0 and m % q == 0:
        print("NO")
        return
    print("YES")
    for i in range(k):
        for j in range(q):
            grid.append(k * q - (i * q + j) - 1)

    grid[k * q - 1] = -(k * q) * (k * q - 1) / 2
    for i in range(n):
        for j in range(m):
            print(int(-grid[(i % k) * q + (j % q)]), end=' ')
        print()
    
    return

main()
